File: 0229-majority-element-ii/0229-majority-element-ii.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        result = []
        elem1, elem2 = 0, 0
        count1, count2 = 0, 0
        n = len(nums)

        for i in range(n): 
            if nums[i] == elem1:
                count1 += 1
            elif nums[i] == elem2:
                count2 += 1
            elif count1 == 0:
                elem1, count1 = nums[i], 1
            elif count2 == 0:
                elem2, count2 = nums[i], 1
            else:
                count1 -= 1
                count2 -= 1
        print(elem1, elem2)
        if nums.count(elem1) > (n // 3):
            result.append(elem1)
        if elem2 is not None and elem1 != elem2 and nums.count(elem2) > (n // 3):
            result.append(elem2)
        return result

File: 0229-majority-element-ii/0229-majority-element-ii.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        int n = nums.length;

        int count1 = 0, count2 = 0;
        int prev1 = Integer.MIN_VALUE, prev2 = Integer.MIN_VALUE;

        for (int i = 0; i < n; i++) {
            if (prev1 == nums[i]) {
                count1++;
                continue;
            } else if (prev2 == nums[i]) {
                count2++;
                continue;
            } else if (count1 == 0) {
                count1++;
                prev1 = nums[i];
                continue;
            } else if (count2 == 0) {
                count2++;
                prev2 = nums[i];
                continue;
            } 
            count1--;
            count2--;
            

        }

        count1 = 0;
        count2 = 0;

        for (int num : nums) {
            if (num == prev1) {
                count1++;
            } else if (num == prev2) {
                count2++;
            }
        }

        List<Integer> result = new ArrayList<>();
        System.out.println(prev1 + " " + prev2);
        if (count1 > n / 3) {
            result.add(prev1);
        }

        if (count2 > n / 3) {
            result.add(prev2);
        }

        return result;
    }
}